src/ws.nim

```diff
@@ -205,226 +280,361 @@ proc encodeFrame(f: Frame): seq[byte] =
-       rsv1: false,
-       rsv2: false,
-       rsv3: false,
-       opcode: opcode,
-       mask: ws.masked,
-       data: data,
-       maskKey: maskKey)
-    var frame = encodeFrame(inFrame)
-    const maxSize = 1024*1024
-    # Send stuff in 1 megabyte chunks to prevent IOErrors.
-    # This really large packets.
-    var i = 0
-    while i < frame.len:
-      let frameSize = min(frame.len, i + maxSize)
-      let res = await ws.tcpSocket.write(frame[i ..< frameSize])
-      if res != frameSize:
-        raise newException(ValueError, "Error while send websocket frame")
-      i += maxSize
-  except OSError, ValueError:
-    # Wrap all exceptions in a WebSocketError so its easy to catch
-    raise newException(WebSocketError, "Failed to send data: " &
-        getCurrentExceptionMsg())
-
-proc sendStr*(ws: WebSocket, data: string, opcode = Opcode.Text): Future[void] =
-  send(ws, toBytes(data), opcode)
-
-proc readFrame(ws: WebSocket): Future[Frame] {.async.} =
-  ## Gets a frame from the WebSocket.
-  ## See https://tools.ietf.org/html/rfc6455#section-5.2
-
-  # Grab the header.
-  var header = newSeq[byte](2)
-  try:
-    await ws.tcpSocket.readExactly(addr header[0], 2)
-  except TransportUseClosedError:
-    ws.readyState = Closed
-    raise newException(WebSocketError, "Socket closed")
-
-  if header.len != 2:
-    ws.readyState = Closed
-    raise newException(WebSocketError, "Invalid websocket header length")
-
-  let b0 = header[0].uint8
-  let b1 = header[1].uint8
+proc send*(
+  ws: WebSocket,
+  data: seq[byte] = @[],
+  opcode = Opcode.Text): Future[void] {.async.} =
+  ## Send a frame
+  ##
+
+  logScope:
+    opcode = opcode
+    dataSize = data.len
+
+  debug "Sending data to remote"
+
+  var maskKey: array[4, char]
+  if ws.masked:
+    maskKey = genMaskKey(ws.rng)
+
+  if opcode notin {Opcode.Text, Opcode.Cont, Opcode.Binary}:
+    discard await ws.tcpSocket.write(encodeFrame(Frame(
+        fin: true,
+        rsv1: false,
+        rsv2: false,
+        rsv3: false,
+        opcode: opcode,
+        mask: ws.masked,
+        data: data, # allow sending data with close messages
+        maskKey: maskKey)))
 
-  var frame: Frame
-  # Read the flags and fin from the header.
+    return
 
-  var hf = cast[HeaderFlags](b0 shr 4)
-  frame.fin = fin in hf
-  frame.rsv1 = rsv1 in hf
-  frame.rsv2 = rsv2 in hf
-  frame.rsv3 = rsv3 in hf
+  let maxSize = ws.frameSize
+  # Send stuff in 1 megabyte chunks to prevent IOErrors.
+  # This really large packets.
+  var i = 0
+  while i < data.len:
+    let len = min(data.len, (maxSize + i))
+    let inFrame = Frame(
+        fin: if (i + len >= data.len): true else: false,
+        rsv1: false,
+        rsv2: false,
+        rsv3: false,
+        opcode: if i > 0: Opcode.Cont else: opcode, # fragments have to be `Continuation` frames
+        mask: ws.masked,
+        data: data[i ..< len],
+        maskKey: maskKey)
+
+    discard await ws.tcpSocket.write(encodeFrame(inFrame))
```

Comment:
Shouldn't we verify the encoded frame length after writing to the socket or have an exception handler around the write?

Reply:
We probably want the error from the write to propagate to the caller, we might want to handle it and close the socket/cleanup and then rethrow tho.

Reply:
For the lenght, as far as I can tell from the spec, it only cares about the payload?